velox/functions/lib/aggregates/WeightedAverageAggregateBase.h

```diff
@@ -0,0 +1,520 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/exec/Aggregate.h"
+#include "velox/functions/lib/aggregates/DecimalAggregate.h"
+#include "velox/type/DecimalUtil.h"
+#include "velox/vector/ComplexVector.h"
+#include "velox/vector/DecodedVector.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::functions::aggregate {
+
+namespace {
+
+/// Translate selected rows of decoded to the corresponding rows of its base
+/// vector.
+SelectivityVector translateToInnerRows(
+    const DecodedVector& decoded,
+    const SelectivityVector& rows) {
+  VELOX_DCHECK(!decoded.isIdentityMapping());
+  if (decoded.isConstantMapping()) {
+    auto constantIndex = decoded.index(rows.begin());
+    SelectivityVector baseRows{constantIndex + 1, false};
+    baseRows.setValid(constantIndex, true);
+    baseRows.updateBounds();
+    return baseRows;
+  } else {
+    SelectivityVector baseRows{decoded.base()->size(), false};
+    rows.applyToSelected(
+        [&](auto row) { baseRows.setValid(decoded.index(row), true); });
+    baseRows.updateBounds();
+    return baseRows;
+  }
+}
+
+/// Return the selected rows of the base vector of decoded corresponding to
+/// rows. If decoded is not identify mapping, baseRowsHolder contains the
+/// selected base rows. Otherwise, baseRowsHolder is unset.
+const SelectivityVector* getBaseRows(
+    const DecodedVector& decoded,
+    const SelectivityVector& rows,
+    SelectivityVector& baseRowsHolder) {
+  const SelectivityVector* baseRows = &rows;
+  if (!decoded.isIdentityMapping() && rows.hasSelections()) {
+    baseRowsHolder = translateToInnerRows(decoded, rows);
+    baseRows = &baseRowsHolder;
+  }
+  return baseRows;
+}
+
+} // namespace
+
+/// Final = SUM(value * weight) / SUM(weight)
+/// To avoid overflow, both TAccumulatorSum and TAccumulatorWeight are +8
+/// bytes(double). Otherwise, we need to add logic to check plus and multiply
+/// overflow.
+/// Decimal type is not supported yet.
+template <typename TAccumulatorSum, typename TAccumulatorWeight>
+struct SumWeight {
+  TAccumulatorSum sum{0};
+  TAccumulatorWeight weight{0};
+};
+
+template <
+    typename TInputValue,
+    typename TInputWeight,
+    typename TAccumulatorSum,
+    typename TAccumulatorWeight,
+    typename TResult>
+class WeightedAverageAggregateBase : public exec::Aggregate {
```

Comment:
is this possible as simple @kagamiori

Reply:
Yes, I believe this function can be implemented using the simple interface. The current implementation has a fast path for constant-encoded inputs that the simple interface cannot provide though. But unless the input batch is very large and inputs are often constant-encoded, this should not affect the performance much. So overall I'd suggest writing the function through the simple interface that would make the implementation much simpler. https://facebookincubator.github.io/velox/develop/aggregate-functions.html#simple-function-interface.